### ISSUE TITLE
[Enhancement]: Block-based checkout - Update address fields display logic

### ISSUE THREAD

**Issue by nielslange** [woothemes/woocommerce#46512]
> [!IMPORTANT]
> Sub issue #46581 needs to be implemented first, as #46582 build on top of that issue.

```[tasklist]
### Tasks
- [ ] #46581
- [ ] #46582
```

### Describe the solution you'd like

This is the parent issue of #46581 and #46582 and aims to update the display logic of the address fields in the Checkout block and the address forms within the My Account section. Detailed descriptions can be found in the corresponding sub issues.

## Screenshot

![CleanShot 2024-04-09 at 11 00 24@2x](https://github.com/woocommerce/woocommerce/assets/3323310/6c8cde08-3fe6-4ddc-92c7-285f6ae1ced6)

### Display logic

- The `Conmany name` and `Phone number` fields are visibly independent if they are optional or required.
- When the `Address line 2` field is set to required, it's visible by default.
- When the `Address line 2` field is set to optional, it's hidden behind a link called `+ Add apartment, suitenumber, etc.`
- When the `Address line 2` field is set to optional, but it contains text, it's visible by default. It remains visible, when the seller empties the field during the checkout. 

### Existing behaviour of the classic checkout when using a classic theme

When using the classic checkout and a classic theme, the `Company name`, `Address line 2` and `Phone number` fields have three options:
- `hidden` → The corresponding field will be hidden on the checkout page and in the address forms within the My Account area.
- `optional` → The corresponding field will be optional on the checkout page and in the address forms within the My Account area. Both the checkout page and in the address forms can be submitted when the corresponding field is empty.
- `required` → The corresponding field will be required on the checkout page and in the address forms within the My Account area. Both the checkout page and in the address forms can only be submitted when the corresponding field is filled out.

<details>
<summary>Screenshots</summary>

#### Hidden fields

<img width="1800" alt="Screenshot 2024-04-13 at 13 19 22" src="https://github.com/woocommerce/woocommerce/assets/3323310/34e514e2-9431-456b-9ae7-e62d1d6e361b">

#### Optional fields

<img width="1800" alt="Screenshot 2024-04-13 at 13 20 05" src="https://github.com/woocommerce/woocommerce/assets/3323310/1d1a15a9-28a6-41bc-bbc7-94c300c2daa1">

#### Required fields

<img width="1800" alt="Screenshot 2024-04-13 at 13 20 35" src="https://github.com/woocommerce/woocommerce/assets/3323310/006e1a0f-13e7-4c09-9f3b-95499d52f597">

<details>

**Comment by nielslange** [comment]
@pmcpinto and @ralucaStan In my initial implementation of the `Address line 2` link, I used the following logic:
- When the `Address line 2` field should be displayed, and it contains text, we're not hiding the input field behind a link, but show the input field directly. When a seller empties that field during the checkout process, the field remains visible, in case the seller wants to fill it with different information.

In the issue description above, I rephrased this part to:
- When the Address line 2 field is set to optional, but it contains text, it's visible by default. It remains visible, when the seller empties the field during the checkout.

Can you confirm that this is how we want to handle such a scenario? On Baymard, I could not find any information regarding this specific case, and I believe that hiding available information behind a link, might lead to accidentally submitting incorrect billing/shipping data.

**Comment by ralucaStan** [comment]
@nikkivias we should have some guidance on how that toggle animation should look.

> When the Address line 2 field is set to optional, but it contains text, it's visible by default. It remains visible, when the seller empties the field during the checkout.
Would that impact the block setting? What happens if the field is marked as hidden, but it has a value?

**Comment by pmcpinto** [comment]
> When the Address line 2 field is set to optional, but it contains text, it's visible by default. It remains visible, when the seller empties the field during the checkout.

@nielslange, do you have specific examples of what kind of text users could add?

**Comment by nielslange** [comment]
> @nielslange, do you have specific examples of what kind of text users could add?

Certainly, @pmcpinto. For instance, in my current residence, a housing complex with two gates, all neighbors share the same street address, but have different unit numbers, such as **Unit B15**. Previously, I lived in a complex with three residential towers, where each apartment had a unique number, like **Unit 17A4**. This number indicates that it's apartment number 4 on the 17th floor of Tower A.

Therefore, a user might enter **Unit B15** or **Unit 17A4** in the address field to specify their exact location. Having the field hidden behind a link, even when it's filled out, could lead to shipments being sent to the wrong address or not being deliverable at all. Therefore, I believe it's important to not collapse this field, even when marked as optional, when it's filled.

**Comment by dominique-vias** [comment]
> When the Address line 2 field should be displayed, and it contains text,

> Having the field hidden behind a link, even when it's filled out


I'm a little confused by this @nielslange

The Address Line 2 field is hidden behind a link by default, however as soon as they click the link to reveal the field, there really isn't a way to revert it back to a link— unless the field is empty and they refresh the page. If they click the link, and enter text into the field, it should remain visible.

**Comment by nielslange** [comment]
> The Address Line 2 field is hidden behind a link by default, however as soon as they click the link to reveal the field, there really isn't a way to revert it back to a link— unless the field is empty and they refresh the page. If they click the link, and enter text into the field, it should remain visible.

Sorry in case I didn't communicate clear enough, @nikkivias. Let's say the `Address line 2` field is set to optional. When the shopper never filled out the `Address line 2` field, it will be hidden behind the `+ Add apartment, suite, etc.` link:

<img width="694" alt="Screenshot 2024-04-17 at 11 01 00" src="https://github.com/woocommerce/woocommerce/assets/3323310/42c67d51-2b2d-4a21-9d9b-461279cad815">

If the shopper filled out the `Address line 2` field in the past and now placing another order, the `Address line 2` field is not hidden behind the link, but visible immediately because it is pre-filled already:

<img width="714" alt="Screenshot 2024-04-17 at 11 02 10" src="https://github.com/woocommerce/woocommerce/assets/3323310/a56a438c-67cb-45d9-9aaa-349b7f5851b6">

The customer's billing and shipping address are stored in their user profile, e.g. `/wp-admin/profile.php`, which shows your user profile. My suggestion is, if `Address line 2` is filled in the user profile, we should not hide this field behind a link in the Checkout block. Does that make more sense now?

**Comment by dominique-vias** [comment]
I see what you mean now, thanks @nielslange This makes perfect sense.

If a shopper is returning and the `Address Line 2` from a saved address is filled, it makes perfect sense that it should be visible by default. This should also apply to any shipping fields that were filled out from a previous transaction.

**Comment by nielslange** [comment]
Thanks for confirming this behaviour, @nikkivias. 🙌

**Comment by pmcpinto** [comment]
@nielslange thanks for the clarification. If it's a returning customer, he should see the returning shipping card:

<img width="621" alt="Captura de ecrã 2024-04-19, às 16 06 16" src="https://github.com/woocommerce/woocommerce/assets/3380168/ad8af036-4b57-447e-9c7f-b0257aac4ecb">

Or are you talking about when the user clicks to edit the card?

**Comment by nielslange** [comment]
> @nielslange thanks for the clarification. If it's a returning customer, he should see the returning shipping card:
> [...]
> Or are you talking about when the user clicks to edit the card?

Thanks for asking, @pmcpinto. I was talking about the billing and shipping forms, not the address card.

**Comment by pmcpinto** [comment]
@opr who can follow-up on this on the Rubik side while Niels is away?

**Comment by opr** [comment]
@pmcpinto looks like #46582 is blocked, but we may be able to be unblock it by using entities to solve the problem. Are you proposing we add this to cycle 8? @samueljseay or @alexflorisca may be able to pick this up in that case.

**Comment by opr** [comment]
@pmcpinto after looking at it more, I see why it's blocked. Entities likely _won't_ solve this issue, as that implies all checkout blocks will be configured the same. The question is, if a site has multiple checkout blocks, which one does the My Account area use as the source of truth? If we can answer that then we can work on the issue probably.

**Comment by pmcpinto** [comment]
@opr, thanks for the clarification, and sorry for the late reply. IMO, asking the user to identify which block can be used as the source of truth is overkill for the first iteration. A good starting point is to use the Checkout block displayed on the Checkout page as the source of truth. It's not perfect, but it will cover most use cases. Do you think it's doable?

**Comment by samueljseay** [comment]
I'm curious about this, it's interesting to even contemplate why we'd have block level settings for address fields being optional/required. Does that really make sense for 99% of merchants? And this issue where we want to "globally" match address validation to a single checkout block makes this more prominent imo. Why don't we move these settings out of the block itself? Could be a naive question.

I could imagine that we could link to address settings from the block to ensure that users can still configure them easily.

**Comment by ralucaStan** [comment]
The baked-in settings at the block level were historically seen as an advantage that the blocks offered: more control and ease of setup. 

We discussed the balance between what needs to be at block level and what needs to live in the WooCommerce Admin on multiple occasions without a clear answer. There were some discussions about differentiating between global vs local pca54o-y3-p2#comment-2490 and even about a separate C&C admin area p6q8Tx-1mY-p2. It was always something that got pushed because of other priorities. 

There are other [issues around settings](https://github.com/woocommerce/woocommerce/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22focus%3A%20settings%22%20label%3A%22team%3A%20Rubik%22), so if we do move address settings it needs to be based on a decision pattern that applies to other areas as well. 

Might be a good time to restart that conversation, but I think this will need design support as well for analysis and agreement on an approach.

**Comment by pmcpinto** [comment]
> I'm curious about this, it's interesting to even contemplate why we'd have block level settings for address fields being optional/required. Does that really make sense for 99% of merchants? And this issue where we want to "globally" match address validation to a single checkout block makes this more prominent imo. Why don't we move these settings out of the block itself? Could be a naive question.

This is a good question. My POV on this has changed a bit over the years. IMO we should have a clear and simple mental model for the checkout-related settings:
- Functional behavior > admin settings
- Visual changes > editor settings

Part of the challenge is that some products (eg: payments) leveraging the checkout surface area have a hybrid approach, as they also include visual updates in the admin settings. I'm not 100% confident that this fragmentation can be avoided, considering these products  need/want to allow users to easily make some visual changes regardless of the checkout used. 
Anyway, at least for the Core settings I think we should follow what I proposed above. But, we should wait until we have a designer working with us, to guarantee that we have alignment between all the functions and this is the best long-term solution.

**Comment by senadir** [comment]
Second part of issue here (apply settings to my account) was solved with @mikejolley entities work, should we close this or does it still need design?

**Comment by pmcpinto** [comment]
> Second part of issue here (apply settings to my account) was solved with @mikejolley entities work

Do you mean this one? #46582

**Comment by senadir** [comment]
Yes! I closed that issue for now because what's written in it is already solved.

**Comment by ralucaStan** [comment]
@pmcpinto can you check the current implementation after https://github.com/woocommerce/woocommerce/issues/46582? I don't have context about why this needs design changes. 

If there are changes you want for the current implementation I suggest creating a new ticket and closing this one.

**Comment by pmcpinto** [comment]
> Yes! I closed that issue for now because what's written in it is already solved.

Thanks for the confirmation. 

> @pmcpinto can you check the current implementation after https://github.com/woocommerce/woocommerce/issues/46582? I don't have context about why this needs design changes.
> If there are changes you want for the current implementation I suggest creating a new ticket and closing this one.

The needs:design label was mainly for discussing admin vs editor settings. But this topic is too large to be discussed in this ticket, so I'll close it. Sorry for the confusion.
